pkgs/top-level/all-packages.nix

```diff
@@ -31055,6 +31055,12 @@ with pkgs;
 
   cordless = callPackage ../applications/networking/instant-messengers/cordless { };
 
+  cosmic-comp = callPackage ../applications/window-managers/cosmic/comp { };
+
+  cosmic-applets = callPackage ../applications/window-managers/cosmic/applets { };
+
+  cosmic-panel = callPackage ../applications/window-managers/cosmic/panel { };
```

Comment:
Is there a reason why `cosmic-panel` and `cosmic-applets` don't use `buildRustPackage` like `cosmic-comp` does?

There's a lot of important logic in `buildRustPackage` for configuring `cargo` correctly, especially for cross compilation.  Using `stdenv.mkDerivation` means missing out on that, which is why `pkgsCross.aarch64-multiplatform.cosmic-comp` fails to build that way -- cargo isn't even getting the correct `--target` flag.

With the patch below, and my comments above, I can build `pkgsCross.aarch64-multiplatform.cosmic-{comp,applets,panel}`.

```patch
diff --git a/pkgs/applications/window-managers/cosmic/applets/default.nix b/pkgs/applications/window-managers/cosmic/applets/default.nix
index 6c0d2a04461d..3232729c9963 100644
--- a/pkgs/applications/window-managers/cosmic/applets/default.nix
+++ b/pkgs/applications/window-managers/cosmic/applets/default.nix
@@ -4,7 +4,7 @@
 , util-linux
 }:

-stdenv.mkDerivation {
+rustPlatform.buildRustPackage {
   pname = "cosmic-applets";
   version = "unstable-2023-08-23";

@@ -15,7 +15,7 @@ stdenv.mkDerivation {
     hash = "sha256-gf8xf/vzVw3yg+LpmOR+z7VJQtEoMEJ+CFO2vgMmFtM=";
   };

-  cargoDeps = rustPlatform.importCargoLock {
+  cargoLock = {
     lockFile = ./Cargo.lock;
     outputHashes = {
       "accesskit-0.11.0" = "sha256-/6KUCH1CwMHd5YEMOpAdVeAxpjl9JvrzDA4Xnbd1D9k=";
@@ -39,6 +39,7 @@ stdenv.mkDerivation {
   buildInputs = [ dbus glib libxkbcommon pulseaudio wayland ];

   justFlags = [ "--set" "prefix" (placeholder "out") ];
+  dontUseJustBuild = 1;  # we need to let cargo-setup-hook do this; it knows the correct flags

   # Force linking to libwayland-client, which is always dlopen()ed.
   "CARGO_TARGET_${rust.lib.toRustTargetForUseInEnvVars stdenv.hostPlatform}_RUSTFLAGS" = map (a: "-C link-arg=${a}") [
@@ -47,6 +48,12 @@ stdenv.mkDerivation {
     "-Wl,--pop-state"
   ];

+  # upstream justfile `install` target does not understand non-native compilation
+  preInstall = lib.optionalString (rust.lib.toRustTarget stdenv.buildPlatform != rust.lib.toRustTarget stdenv.hostPlatform) ''
+    rm -rf target/release
+    mv target/${rust.lib.toRustTarget stdenv.hostPlatform}/release target/release
+  '';
+
   passthru.updateScript = unstableGitUpdater { };

   meta = with lib; {
diff --git a/pkgs/applications/window-managers/cosmic/panel/default.nix b/pkgs/applications/window-managers/cosmic/panel/default.nix
index a102504ee139..d598c7e717d8 100644
--- a/pkgs/applications/window-managers/cosmic/panel/default.nix
+++ b/pkgs/applications/window-managers/cosmic/panel/default.nix
@@ -3,7 +3,7 @@
 , unstableGitUpdater
 }:

-stdenv.mkDerivation {
+rustPlatform.buildRustPackage {
   pname = "cosmic-panel";
   version = "unstable-2023-08-03";

@@ -14,7 +14,7 @@ stdenv.mkDerivation {
     hash = "sha256-H3QuiP7Og69wm9yCX/uoSG0aQ3B/61q9Sdj+rW4KZMU=";
   };

-  cargoDeps = rustPlatform.importCargoLock {
+  cargoLock = {
     lockFile = ./Cargo.lock;
     outputHashes = {
       "cosmic-config-0.1.0" = "sha256-ENdQkLo4MSH/oe6vFr8qByWehch0nMngBr9/6hF7iVk=";
@@ -31,6 +31,7 @@ stdenv.mkDerivation {
   buildInputs = [ libglvnd libxkbcommon wayland ];

   justFlags = [ "--set" "prefix" (placeholder "out") ];
+  dontUseJustBuild = 1;  # we need to let cargo-setup-hook do this; it knows the correct flags

   # Force linking to libEGL, which is always dlopen()ed.
   "CARGO_TARGET_${rust.lib.toRustTargetForUseInEnvVars stdenv.hostPlatform}_RUSTFLAGS" = map (a: "-C link-arg=${a}") [
@@ -39,6 +40,12 @@ stdenv.mkDerivation {
     "-Wl,--pop-state"
   ];

+  # upstream justfile `install` target does not understand non-native compilation
+  preInstall = lib.optionalString (rust.lib.toRustTarget stdenv.buildPlatform != rust.lib.toRustTarget stdenv.hostPlatform) ''
+    rm -rf target/release
+    mv target/${rust.lib.toRustTarget stdenv.hostPlatform}/release target/release
+  '';
+
   passthru.updateScript = unstableGitUpdater { };

   meta = with lib; {
```

Reply:
> There's a lot of important logic in `buildRustPackage` for configuring `cargo` correctly, especially for cross compilation. Using `stdenv.mkDerivation` means missing out on that

`buildRustPackage` and other wrappers around `mkDerivation` are horrible and should go away, but you're otherwise correct that not using `cargoBuildHook` means missing out on stuff.  The reason to use just here was that Cargo, not being a proper build system, lacks a concept of installation, and so our options for installing the binary and supporting files are either using `just install`, or duplicating the install logic ourselves.

Is there still important logic you think we'd be missing out on if I fixed the justfile to understand cross targets?

Reply:
> `buildRustPackage` and other wrappers around `mkDerivation` are horrible and should go away (because they don't compose and overriding them doesn't work)

I agree, but we need a `rustc-wrapper` in order to do that.

Using `just` for the `installPhase` totally makes sense.

> Is there still important logic you think we'd be missing out on if I fixed the justfile to understand cross targets?

Yeah, the stuff in `cargo-build-hook`:

https://github.com/NixOS/nixpkgs/blob/8c9fc0efefedf6e4a262ccf4d9eba9df7e5ef743/pkgs/build-support/rust/hooks/cargo-build-hook.sh#L31-L45

... which falls into two categories:

1. The environment variables
2. The flags

The other issue is that because we don't have a `NIX_RUSTC_FLAGS_{COMPILE,LINK}`, 
there isn't (currently) a way for you to reference this set of environment variables and flags (https://github.com/NixOS/nixpkgs/pull/247442 fixes this for env vars but not for flags) other than by letting `buildRustPackage` drive the build.